### PR TITLE
Ensure subclasses inherit fields

### LINKED ISSUE #6: Subclasses should return inherited fields in #to_h
If I create a class with some fields in and I then subclass it, the `#to_h` method on the subclass should return the fields from the parent class as well as any of its own.

### PR DESCRIPTION
**Pull request by chrismytton**
This adds an `inherited` hook to ensure that subclasses inherit fields correctly.

Fixes #6

### CODE REVIEW

File: test/field_serializer_test.rb [hunk 1]
```diff
@@ -68,4 +68,24 @@ class InternalFieldTest
...
+      field :state do
+        'Zamfara'
+      end
+    end
+
+    it 'returns fields from the parent class in #to_h' do
+      InheritanceSubclassTest.new.to_h.must_equal(constituency: 'Bungudu', state: 'Zamfara')
+    end
+  end
 end
```
Comment (tmtmtmtm) [inline test/field_serializer_test.rb]: I think you also want a test here to show that it DTRT in cases where a field exists in both the parent and the subclass…

**Comment by ondenman** [comment]
@tmtmtmtm Added test to show that field in subclass overrides field in parent class.

**Comment by chrismytton** [comment]
@ondenman tl;dr I think this can probably be done in a simpler way using Ruby's meta-programming tools. I've written in a bit more detail below to hopefully give you some insight into which tools in particular might be suitable for this job and why.

---

As you may know, Ruby has "class variables", "instance variables" and "class instance variables". The current implementation of `FieldSerializer` (on the `master` branch) uses "class instance variables". This is why fields aren't currently being inherited by subclasses.

My initial thought was that this _might_ be fixable by simply switching to "class variables".

``` diff
diff --git a/lib/field_serializer.rb b/lib/field_serializer.rb
index 2f493dd..98e9af8 100644
--- a/lib/field_serializer.rb
+++ b/lib/field_serializer.rb
@@ -12,7 +12,7 @@ module FieldSerializer
   module ClassMethods

     def fields
-      @fields ||= []
+      @@fields ||= []
     end

     def field(name, &block)
```

This doesn't work! 😨  

It turns out (and this was new to me!) that if you use class variables in a module such as `FieldSerializer`, then the `@@fields` class variable is defined _on the `FieldSerializer` module_, not on the class the module is mixed into.

We can fix this problem with Ruby's metaprogramming tools. Using `class_variable_defined?`, `class_variable_set` and `class_variable_get` <sup>[1](#fn-1)</sup> allows us to set and get "class variables" on the class that `FieldSerializer` is mixed into, rather than on the module.

``` diff
diff --git a/lib/field_serializer.rb b/lib/field_serializer.rb
index 2f493dd..caac978 100644
--- a/lib/field_serializer.rb
+++ b/lib/field_serializer.rb
@@ -12,7 +12,11 @@ module FieldSerializer
   module ClassMethods

     def fields
-      @fields ||= []
+      if class_variable_defined?(:@@fields)
+        class_variable_get(:@@fields)
+      else
+        class_variable_set(:@@fields, [])
+      end
     end

     def field(name, &block)
```

This works because `class_variable_defined?` and friends are being called on `self`, which at that point will be the target class.

---

<a name="fn-1">1</a>: [This blog post explains these in a bit more detail.](http://thoughts.codegram.com/understanding-class-instance-variables-in-ruby/)

**Comment by tmtmtmtm** [comment]
@chrismytton I think this approach will crash horribly if something then further subclasses a subclass, e.g.

``` patch
diff --git a/test/field_serializer_test.rb b/test/field_serializer_test.rb
index 1cf43b5..ba082c7 100644
--- a/test/field_serializer_test.rb
+++ b/test/field_serializer_test.rb
@@ -94,6 +94,20 @@ describe FieldSerializer  do
       end
     end

+    class InheritanceSubclassTest3 < InheritanceSubclassTest2
+      field :constituency do
+        'Wibble'
+      end
+
+      field :district do
+        'Bar'
+      end
+    end
+
+    it 'should allow subclasses to be subclassed' do
+      InheritanceSubclassTest3.new.to_h.must_equal(constituency: 'Wibble', state: 'Zamfara', district: 'Bar')
+    end
+
     it 'should return fields defined by parent class in #to_h' do
       InheritanceSubclassTest.new.to_h.must_equal(constituency: 'Bungudu', state: 'Zamfara')
```

Because of the way class variables are shared, now the parent classes of `InheritanceSubclassTest3` all believe they should have a `:district` field, so `InheritanceSubclassTest2` blows up with a `undefined method` error.

**Comment by chrismytton** [comment]
@tmtmtmtm 😱  You're quite right! Hmm, in that case I'm not sure what the best solution is here. It certainly seems like something that _should_ be possible without having to jump through hoops calling `super`, but nothing is jumping to mind immediately.

**Comment by chrismytton** [comment]
I've pushed a couple of commits which make this change a bit simpler by using Ruby's `inherited` hook. I've also added the test that @tmtmtmtm suggested.

**Comment by chrismytton** [comment]
@tmtmtmtm 👀

**Comment by tmtmtmtm** [comment]
Woot! Much much nicer now! I was going to say that this probably needs an entry in the Changelog, but it seems there isn't one. 

We probably should update the README though, as this is a fairly significant feature. (Though I'm OK with this being merged now and the README added separately if this is blocking something)

**Review by tmtmtmtm — APPROVED** [review]

**Pull request merged**